vscode/quint-vscode/server/src/complete.ts

```diff
@@ -0,0 +1,250 @@
+import { compact } from 'lodash'
+import { MarkupContent } from 'vscode-languageclient'
+import { CompletionItem, CompletionItemKind, MarkupKind, Position } from 'vscode-languageserver/node'
+import {
+  DocumentationEntry,
+  IRVisitor,
+  ParserPhase3,
+  QuintDeclaration,
+  QuintDef,
+  QuintEx,
+  QuintLambda,
+  QuintModule,
+  QuintType,
+  analyzeModules,
+  booleanOperators,
+  findDefinitionWithId,
+  findExpressionWithId,
+  integerOperators,
+  listOperators,
+  mapOperators,
+  recordOperators,
+  setOperators,
+  tupleOperators,
+  typeToString,
+  walkDeclaration,
+  walkExpression,
+} from '@informalsystems/quint'
+import { findMatchingResults } from './reporting'
+
+export function completeIdentifier(
+  identifier: string,
+  parsedData: ParserPhase3,
+  sourceFile: string,
+  position: Position,
+  builtinDocs: Map<string, DocumentationEntry>
+): CompletionItem[] {
+  const { modules, table } = parsedData
+
+  // Until we have incremental parsing, try to lookup `triggeringIdentifier` by name
+  const declIds = findDeclByNameAtPos(identifier, position, sourceFile, parsedData)
+
+  // FIXME: use analysisOutputByDocument
+  const [_, aop] = analyzeModules(table, modules)
```

Comment:
Have you tried using `analysisOutputByDocument`? I'm not sure exactly what the consequences of the delay might be, but we might be fine with it.

We might also be fine with running the analysis when users type `.`. The reason for the delay is that we can't run it every time a user types anything, since that becomes laggy pretty quickly.

Reply:
We could try to trigger the async job here and avoid running it twice, but I'm not sure it's worth the trouble. The real improvement will be with incremental parsing.

Reply:
> Have you tried using analysisOutputByDocument? I'm not sure exactly what the consequences of the delay might be, but we might be fine with it.

Yeah, I tried that, but in 99% of cases the analysis output and the parsed data are out of sync by then (i.e., using disjoint sets of IDs).

I guess the delay is fine for hover, because an analysis will be triggered until you move your mouse, but for triggering completion, the LSP request wins the race against the timeout.

Reply:
Maybe we should make `analysisOutputByDocument` a Promise? Then I could wait for that.

Reply:
I tried to combine this with a promise, but I couldn't find a way to use the same analysis action both with a timeout and as a promise.

I'm off next week, maybe you have an idea here, otherwise I'll take another stab at it when I'm back!

Reply:
I gave this a shot and I think I got something pretty OK, please take a look and see if you  agree: https://github.com/informalsystems/quint/pull/1143/commits/ea330e09bfef2ef64de825d489acac3d7f6afd19

Reply:
This looks great! Thanks for taking a shot at it! 🙏